client/modules/IDE/components/Editor.jsx

```diff
@@ -155,14 +160,21 @@ class Editor extends React.Component {
       }
     }, 1000));
 
-    this._cm.on('keyup', () => {
+    this.map = {};
+
+    this._cm.on('keyup', (_cm, e) => {
       const temp = this.props.t('Editor.KeyUpLineNumber', { lineNumber: parseInt((this._cm.getCursor().line) + 1, 10) });
       document.getElementById('current-line').innerHTML = temp;
+      this.handleKey(this.map, e);
     });
 
     this._cm.on('keydown', (_cm, e) => {
-      // 9 === Tab
-      if (e.keyCode === 9 && e.shiftKey) {
+      this.handleKey(this.map, e);
```

Comment:
I'm not sure how using this function `this.handleKey` is helping in this scenario? Just from reading this it seems like it's adding some intermediate code that is unnecessary, but please explain if I'm wrong!

Reply:
The reason why I use this function is because to detect shortcut <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>F</kbd> the following conditional statement would not work
```
if (e.keyCode === 91 && e.keyCode === 16 && e.keyCode === 70){ }
```
I guess the reason is because the browser cannot detect multiple key press. To solve this, I used an object called `this.map` to store the `keydown` state of the keyboard. For example, if the shortcut <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>F</kbd> is pressed, `this.map` will be
```
{91: true, 16: true, 70: true}
```
 and when we release the shortcut, `this.map` will be
```
{91: false, 16: false, 70: false}
```
I wonder if there is a better way to get around this.

Reply:
there is! if you look at the documentation for a [KeyboardEvent](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent), you'll see that it contains `metaKey`, which will be true if <kbd>Command</kbd> is pressed, `ctrlKey`, which will be true if <kbd>Control</kbd> is pressed, and `shiftKey`, which will be true if <kbd>Shift</kbd> is pressed.